tests/test_cli_compile.py

```diff
@@ -2702,9 +2706,227 @@ def test_all_extras_fail_with_extra(fake_dists, runner, make_module, fname, cont
     assert exp in out.stderr
 
 
+def _mock_resolver_cls(monkeypatch: MonkeyPatch) -> MagicMock:
+    obj = MagicMock()
+    obj.resolve = MagicMock(return_value=set())
+    obj.resolve_hashes = MagicMock(return_value=dict())
+    cls = MagicMock(return_value=obj)
+
+    monkeypatch.setattr("piptools.scripts.compile.BacktrackingResolver", cls)
+    monkeypatch.setattr("piptools.scripts.compile.LegacyResolver", cls)
+
+    return cls
+
+
+def _mock_build_project_metadata(monkeypatch: MonkeyPatch) -> MagicMock:
+    func = MagicMock(
+        return_value=ProjectMetadata(
+            extras=("e",),
+            requirements=(
+                install_req_from_line("rdep0"),
+                install_req_from_line("rdep1; extra=='e'"),
+            ),
+            build_requirements=(install_req_from_line("bdep0"),),
+        )
+    )
+
+    monkeypatch.setattr("piptools.scripts.compile.build_project_metadata", func)
+
+    return func
+
+
+@backtracking_resolver_only
+@pytest.mark.network
+def test_all_extras_and_all_build_deps(
+    fake_dists_with_build_deps,
+    runner,
+    tmp_path,
+    monkeypatch,
+    current_resolver,
+):
+    """
+    Test that trying to lock all dependencies gives the expected output.
+    """
+    src_pkg_path = os.path.join(PACKAGES_PATH, "small_fake_with_build_deps")
+    # When used as argument to the runner it is not passed to pip
+    monkeypatch.setenv("PIP_FIND_LINKS", fake_dists_with_build_deps)
+
+    with runner.isolated_filesystem(tmp_path) as tmp_pkg_path:
+        shutil.copytree(src_pkg_path, tmp_pkg_path, dirs_exist_ok=True)
+        out = runner.invoke(
+            cli,
+            [
+                "--allow-unsafe",
+                "--output-file",
+                "-",
+                "--quiet",
+                "--no-emit-options",
+                "--no-header",
+                "--all-extras",
+                "--all-build-deps",
+            ],
+        )
+
+    # Note that the build dependencies of our build dependencies are not resolved.
+    # This means that if our build dependencies are not available as wheels then we will not get reproducible results.
+    assert out.exit_code == 0
```

Comment:
I suggest checking for this explicitly. Both as a visualization of the comment and a regression check:
```suggestion
    assert out.exit_code == 0
    # Note that the build dependencies of our build dependencies are not resolved.
    # This means that if our build dependencies are not available as wheels then we will not get reproducible results.
    assert "fake_transient_build_dep" not in out.stdout
```

Reply:
Marking this as resolved.